### PR TITLE
Don't allocate arrays when copying streams

### LINKED ISSUE #1509: [FEATURE] Reduce heap allocation on SKManagedStream.OnReadManagedStream
I've found a big heap allocation on images decode at `SKManagedStream.OnReadManagedStream`. 
Screenshot 2020-09-25 at 23 50 48
Maybe it would be better to use a 81920 bytes length buffer to copy a managed stream data to unmanaged pointer? I hope Marshal.Copy cost is not huge.
Screenshot 2020-09-25 at 23 54 03

### PR DESCRIPTION
**Pull request by mattleibow**
**Description of Change**

Don't allocate arrays when copying streams

**Bugs Fixed**

- Fixes #1509

**API Changes**

None.

**Behavioral Changes**

Reading from a wrapped managed stream will no longer allocate huge arrays. or any arrays. Using a native copy with array pools.

**PR Checklist**

- [ ] Has tests (if omitted, state reason in description)
- [ ] Rebased on top of master at time of PR
- [ ] Changes adhere to coding standard
- [ ] Updated documentation

### CODE REVIEW

**Comment by mattleibow** [comment]
It is a pity that we have to copy twice, first to the managed array and then into the destination pointer.

File: binding/Binding/SKManagedStream.cs [hunk 1]
```diff
@@ -90,18 +87,25 @@ protected override void DisposeManaged ()
...
+			if (buffer == IntPtr.Zero)
+				throw new ArgumentNullException (nameof (buffer));
+			if ((int)size < 0)
+				throw new ArgumentOutOfRangeException (nameof (size));
+
+			if (size == IntPtr.Zero)
+				return IntPtr.Zero;
+
+			using var managedBuffer = Utils.RentArray<byte> ((int)size);
+			var len = stream.Read (managedBuffer.Array, 0, managedBuffer.Length);
```
Comment (Ziriax) [inline binding/Binding/SKManagedStream.cs:99]: Shouldn't this be `var len = stream.Read (managedBuffer.Array, 0, size)`;?

The `managedBuffer.Length` is always the first power of two larger than `size` I believe

Comment (mattleibow) [reply binding/Binding/SKManagedStream.cs:99]: That would be the case if I was using `ArrayPool` directly. I am using my Utils version that makes sure to set the `Length` of the "wrapper" to what I requested.

Comment (mattleibow) [reply binding/Binding/SKManagedStream.cs:99]: Thanks for the review!

Comment (Ziriax) [reply binding/Binding/SKManagedStream.cs:99]: Oh right! I actually remember you making this utils method, I forgot about that.

**Pull request merged**